File: frontend/src/types/history.ts
// PARTPILOT:SYSTEM_HISTORY_TYPES:V408

export type HistoryKind = "audit" | "stock_movement";
export type HistoryJson = Record<string, unknown> | unknown[] | null;

export interface HistoryEntry {
  key: string;
  kind: HistoryKind;
  event_type: string;
  occurred_at: string;
  summary: string | null;

  entity_type: string | null;
  entity_id: number | null;
  entity_label: string | null;

  actor_type: string | null;
  actor_user_id: number | null;
  actor_display_name: string | null;

  part_id: number | null;
  part_number: string | null;
  part_name: string | null;

  reservation_id: number | null;
  reservation_label: string | null;
  project_id: number | null;
  project_label: string | null;

  movement_type: string | null;
  quantity: number | null;
  quantity_delta: number | null;
  quantity_before: number | null;
  quantity_after: number | null;
  reserved_quantity_before: number | null;
  reserved_quantity_after: number | null;
  available_quantity_before: number | null;
  available_quantity_after: number | null;
  unit_price_snapshot: string | null;
  currency_snapshot: string | null;
  reason: string | null;
  note: string | null;
  source: string | null;

  before_json: HistoryJson;
  after_json: HistoryJson;
  metadata_json: HistoryJson;
}

export interface HistoryCollection {
  total: number;
  limit: number;
  offset: number;
  entries: HistoryEntry[];
}

export interface HistoryFacetValue {
  value: string;
  count: number;
}

export interface HistoryActorOption {
  user_id: number;
  display_name: string;
  count: number;
}

export interface HistoryFilterOptions {
  kinds: HistoryFacetValue[];
  entity_types: HistoryFacetValue[];
  event_types: HistoryFacetValue[];
  actor_types: HistoryFacetValue[];
  movement_types: HistoryFacetValue[];
  sources: HistoryFacetValue[];
  actors: HistoryActorOption[];
  earliest_at: string | null;
  latest_at: string | null;
}

export interface HistoryQueryOptions {
  kind?: HistoryKind;
  entityType?: string;
  eventType?: string;
  actorType?: string;
  actorUserId?: number;
  movementType?: string;
  from?: string;
  to?: string;
  query?: string;
  limit?: number;
  offset?: number;
  signal?: AbortSignal;
}

File: backend/app/schemas/history.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


# PARTPILOT:SYSTEM_HISTORY_SCHEMA:V406
HistoryKind = Literal["audit", "stock_movement"]


class HistoryEntryResponse(BaseModel):
    key: str
    kind: HistoryKind
    event_type: str
    occurred_at: datetime
    summary: str | None = None

    entity_type: str | None = None
    entity_id: int | None = None
    entity_label: str | None = None

    actor_type: str | None = None
    actor_user_id: int | None = None
    actor_display_name: str | None = None

    part_id: int | None = None
    part_number: str | None = None
    part_name: str | None = None

    reservation_id: int | None = None
    reservation_label: str | None = None
    project_id: int | None = None
    project_label: str | None = None

    movement_type: str | None = None
    quantity: int | None = None
    quantity_delta: int | None = None
    quantity_before: int | None = None
    quantity_after: int | None = None
    reserved_quantity_before: int | None = None
    reserved_quantity_after: int | None = None
    available_quantity_before: int | None = None
    available_quantity_after: int | None = None
    unit_price_snapshot: str | None = None
    currency_snapshot: str | None = None
    reason: str | None = None
    note: str | None = None
    source: str | None = None

    before_json: dict | list | None = None
    after_json: dict | list | None = None
    metadata_json: dict | list | None = None


class HistoryCollectionResponse(BaseModel):
    total: int = Field(ge=0)
    limit: int = Field(ge=1, le=100)
    offset: int = Field(ge=0)
    entries: list[HistoryEntryResponse]


class HistoryFacetValueResponse(BaseModel):
    value: str
    count: int = Field(ge=0)


class HistoryActorOptionResponse(BaseModel):
    user_id: int
    display_name: str
    count: int = Field(ge=0)


class HistoryFilterOptionsResponse(BaseModel):
    kinds: list[HistoryFacetValueResponse]
    entity_types: list[HistoryFacetValueResponse]
    event_types: list[HistoryFacetValueResponse]
    actor_types: list[HistoryFacetValueResponse]
    movement_types: list[HistoryFacetValueResponse]
    sources: list[HistoryFacetValueResponse]
    actors: list[HistoryActorOptionResponse]
    earliest_at: datetime | None = None
    latest_at: datetime | None = None
